### ISSUE TITLE
Refactoring decisions

### ISSUE THREAD

**Issue by pnrobinson** [monarch-initiative/N2V#115]
Hi everybody,
there are a number of things that need to be refactored. I am going to summarize them here as a list so that we can discuss them if anything winds up (Deities forbidding) not working!

```
self.num_sentences: int = len(self.data)
```
self.data used to be a list but now it is a tf tf.data.Dataset. The following code gets us to an integer
```
c = tf.data.experimental.cardinality(dataset) # c is a tensor
d = c.numpy() # d is a numpy int64
e = int(d)
```
Therefore, let's go for
```
self.num_sentences: int = int(tf.data.experimental.cardinality(self.data))
```
This will work for now but might break if we ever decide to pass a data generator. I am not convinced that we need to know the number of sentences, since we are going to refactor the way words are extracted from sentences anyway.

**Comment by callahantiff** [comment]
Looks good to me. I had also created a function in the `tf_utils` script that might work. It is designed to work for a ragged tensor or `tf.Dataset`:

```python

def gets_tensor_length(tensor: Union[tf.data.Dataset, tf.RaggedTensor]) -> int:
    """Returns the length of a tensorflow object.

    Args:
        tensor: A tf.Dataset or tf.RaggedTensor object.

    Returns:
        An integer representing the length of the tensor.
    """

    if isinstance(tensor, tf.RaggedTensor):
        return tensor.shape[0]
    else:
        return len([edge for edge in tensor])
```

**Comment by pnrobinson** [comment]
The training loop probably needs some refactoring love. This is an excerpt of what we have now:
```
 def train(self) -> None:
        # run training for the given number of steps.
        with trange(1, self.num_steps + 1) as pbar:
            for step in pbar:
                if self.list_of_lists:
                    walkcount = 2
                    batch_x, batch_y = self.next_batch_from_list_of_lists(walkcount, self.num_skips, self.skip_window)
                else:
                    batch_x, batch_y = self.next_batch(self.data, self.batch_size, self.num_skips, self.skip_window)
                self.run_optimization(batch_x, batch_y)
```

But instead we could do something like this
```
def train(self) -> None:
 ## THE FOLLOWING IS ONE EPOCH, AND WE WILL LOOP n_epoch times over it
  if words:
    datawindows = tensor_data.window(size=batch_size)
    for wdw in datawindows:
      lst = list(wdw.as_numpy_iterator()) 
      if len(lst) < window_len:
	print("Warning: Skipping the final %d words of the text" % len(lst))
        continue
      batch_x, batch_y = self.next_batch(lst)
      self.run_optimization(batch_x, batch_y)
   else: # i.e., list of lists
     # in this case, self.data is a RaggedTensor. We can iterate easily
     for sentence in self.data:
       lst = sentence.numpy() #Sentence was a Tensor
       batch_x, batch_y = self.next_batch(lst)
       self.run_optimization(batch_x, batch_y)
```
(Roughly)...we just need to add the "display" code somewhere, and we also want to capture the history of the loss something like a standard keras model does. But since lst is a simple python array in both cases, the get-next-batch  function can be the same now for both list and list-of-list

**Comment by pnrobinson** [comment]
Closed by https://github.com/monarch-initiative/N2V/issues/127
